include/execution.hpp

```diff
@@ -615,18 +649,28 @@ namespace std::execution {
     }
 
     inline constexpr struct as_awaitable_t {
-      template <__single_typed_sender S, class Promise>
-      auto operator()(S&& s, Promise& promise) const
-        noexcept(/*TODO*/ false)
-        -> __impl::__awaitable<__id_t<Promise>, __id_t<remove_cvref_t<S>>> {
-        auto h = coro::coroutine_handle<Promise>::from_promise(promise);
-        return {(S&&) s, h};
+      // Pass through things that are awaitable in any context:
+      template <__awaitable T>
+      friend T&& tag_invoke(as_awaitable_t, T&& t, __ignore) noexcept {
+        return (T&&) t;
+      }
```

Comment:
We shouldn't be defining this as a `tag_invoke()` overload here.

This has the problem that if someone has defined a `const T&` overload of `as_awaitable()` for their type but then it's invoked with an rvalue and also happens to be generally awaitable then it will prefer the generally awaitable version instead of the custom `as_awaitable()` overload.

You should add in the handling for awaitable types as an extra `if constexpr` check in the logic below so that it is definitively considered only if there are no valid customisations of `as_awaitable()` found.

Reply:
Fixed.